zookeeper/source/zk_dn_client.cc

```diff
@@ -0,0 +1,124 @@
+#ifndef RDFS_ZK_CLIENT_DN_CC
+#define RDFS_ZK_CLIENT_DN_CC
+
+#include "zk_dn_client.h"
+#include "zk_lock.h"
+#include <easylogging++.h>
+
+namespace zkclient{
+
+
+    const std::string ZkClientDn::CLASS_NAME = ": **ZkClientDn** : ";
+
+	ZkClientDn::ZkClientDn(const std::string& ip, const std::string& hostname, const std::string& zkIpAndAddress,
+            const uint32_t ipcPort, const uint32_t xferPort) : ZkClientCommon(zkIpAndAddress) {
+
+        data_node_id = DataNodeId();
+        data_node_id.ip = ip;
+        data_node_id.ipcPort = ipcPort;
+
+        // TODO: Fill in other data_node stats
+        data_node_payload = DataNodePayload();
+        data_node_payload.ipcPort = ipcPort;
+        data_node_payload.xferPort = xferPort;
+
+        registerDataNode();
+		LOG(INFO) << "Registered datanode " + build_datanode_id(data_node_id);
+
+	}
+
+	bool ZkClientDn::blockReceived(uint64_t uuid) {
+		// TODO: store file size
+		int error_code;
+		bool exists;
+
+		LOG(INFO) << "DataNode received a block with UUID " << std::to_string(uuid);
+        std::string id = build_datanode_id(data_node_id);
+
+		// Add acknowledgement
+		ZKLock queue_lock(*zk.get(), WORK_QUEUES + WAIT_FOR_ACK_BACKSLASH + std::to_string(uuid));
+		if (queue_lock.lock() != 0) {
+			LOG(ERROR) << CLASS_NAME <<  "Failed locking on /work_queues/wait_for_acks/<block_uuid> " << error_code;
+		}
+		if (zk->exists(WORK_QUEUES + WAIT_FOR_ACK_BACKSLASH + std::to_string(uuid), exists, error_code)) {
+			if (exists) {
+				if(zk->create(WORK_QUEUES + WAIT_FOR_ACK_BACKSLASH + std::to_string(uuid) + "/" + id, ZKWrapper::EMPTY_VECTOR, error_code, true)) {
```

Comment:
I guess it just depends on what we want to do if let's say 3 acks were received, but 1-3 of them disappeared, because the dn's went down. This affects the guarantees about the immediate availability of data during a close. If we "close" a block, do we want to say the block is on 3 nodes (dead or alive) or 3 live nodes, I feel like the first one is an ok guarantee? I think it makes sense to use non-ephemeral to avoid over-replicating too. Should we update the architecture doc?